// @internetarchive/shared-resize-observer/dist/demo/dev/dev-slider.js
import { __decorate } from "tslib";
import { css, html, LitElement } from 'lit';
import { customElement, property, query } from 'lit/decorators.js';
let DevSlider = class DevSlider extends LitElement {
    constructor() {
        super(...arguments);
        this.unit = 'px';
        this.min = 0;
        this.max = 100;
        this.step = 1;
        this.value = 1;
    }
    valueChanged() {
        this.value = parseFloat(this.slider.value);
        const event = new CustomEvent('valueChanged', {
            detail: {
                value: this.value,
            },
        });
        this.dispatchEvent(event);
    }
    get displayValue() {
        return `${this.value}${this.unit}`;
    }
    render() {
        return html `
      <label for="slider">${this.label}</label>
      <input
        type="range"
        min=${this.min}
        max=${this.max}
        step=${this.step}
        value=${this.value}
        id="slider"
        @input=${this.valueChanged}
      />
      <span id="value">${this.displayValue}</span>
    `;
    }
    static get styles() {
        const labelWidth = css `var(--dev-label-width, 15rem)`;
        const sliderWidth = css `var(--dev-slider-width, 25rem)`;
        return css `
      label {
        display: inline-block;
        width: ${labelWidth};
      }

      input[type='range'] {
        width: ${sliderWidth};
      }
    `;
    }
};
__decorate([
    property()
], DevSlider.prototype, "label", void 0);
__decorate([
    property()
], DevSlider.prototype, "unit", void 0);
__decorate([
    property({ type: Number })
], DevSlider.prototype, "min", void 0);
__decorate([
    property({ type: Number })
], DevSlider.prototype, "max", void 0);
__decorate([
    property({ type: Number })
], DevSlider.prototype, "step", void 0);
__decorate([
    property({ type: Number })
], DevSlider.prototype, "value", void 0);
__decorate([
    query('#slider')
], DevSlider.prototype, "slider", void 0);
DevSlider = __decorate([
    customElement('dev-slider')
], DevSlider);
export { DevSlider };
