// @internetarchive/ia-item-navigator/dist/src/your-webcomponent.js
import { __decorate } from "tslib";
import { html, css, LitElement, property, customElement } from 'lit-element';
let YourWebComponent = class YourWebComponent extends LitElement {
    constructor() {
        super(...arguments);
        this.title = 'Hey there';
        this.counter = 5;
    }
    increment() {
        this.counter += 1;
    }
    render() {
        return html `
      <h2>${this.title}, Number: ${this.counter}!</h2>
      <button @click=${this.increment}>increment</button>
      <slot name="my-slot"> </slot>
    `;
    }
};
YourWebComponent.styles = css `
    :host {
      display: block;
      padding: 25px;
      color: var(--your-webcomponent-text-color, #000);
    }
  `;
__decorate([
    property({ type: String })
], YourWebComponent.prototype, "title", void 0);
__decorate([
    property({ type: Number })
], YourWebComponent.prototype, "counter", void 0);
YourWebComponent = __decorate([
    customElement('your-webcomponent')
], YourWebComponent);
export { YourWebComponent };
