// @internetarchive/ia-item-navigator/dist/src/item-inspector/grouped-files/files-by-type-provider.js
import { html } from 'lit-element';
import '@internetarchive/ia-sharing-options';
import '@internetarchive/icon-share/icon-share';
export class GroupedFilesProvider {
    constructor(shareArgs) {
        this.item = null;
        this.baseHost = 'https://archive.org';
        this.icon = html `<ia-icon-share style="width: var(--iconWidth); height: var(--iconHeight);"></ia-icon-share>`;
        this.id = 'share';
        this.label = 'Share this item';
        this.component = null;
        this.subPrefix = '';
        this.encodedSubPrefix = '';
        const { item, baseHost, subPrefix = '' } = shareArgs;
        const { identifier = '', creator = '', title = '' } = item.metadata || {};
        const encodedSubPrefix = encodeURIComponent(subPrefix);
        const urlIdentifier = subPrefix && subPrefix !== identifier
            ? `${identifier}/${encodedSubPrefix}`
            : identifier;
        this.item = item;
        this.baseHost = baseHost;
        this.subPrefix = subPrefix;
        this.encodedSubPrefix = encodedSubPrefix;
        this.component = html `<ia-sharing-options
      identifier=${urlIdentifier}
      type="item"
      creator="${creator}"
      description="${title}"
      baseHost="${this.baseHost}"
    ></ia-sharing-options>`;
    }
}
